File: domain/src/main/java/com/arturogutierrez/openticator/domain/account/interactor/AddAccountInteractor.java
package com.arturogutierrez.openticator.domain.account.interactor;

import com.arturogutierrez.openticator.domain.account.AccountFactory;
import com.arturogutierrez.openticator.domain.account.model.Account;
import com.arturogutierrez.openticator.domain.account.repository.AccountRepository;
import com.arturogutierrez.openticator.domain.category.CategoryFactory;
import com.arturogutierrez.openticator.domain.category.CategorySelector;
import com.arturogutierrez.openticator.domain.category.model.Category;
import com.arturogutierrez.openticator.domain.category.repository.CategoryRepository;
import com.arturogutierrez.openticator.executor.PostExecutionThread;
import com.arturogutierrez.openticator.executor.ThreadExecutor;
import com.arturogutierrez.openticator.interactor.Interactor;
import rx.Observable;

public class AddAccountInteractor extends Interactor<Account> {

  private final AccountRepository accountRepository;
  private final AccountFactory accountFactory;
  private final CategoryRepository categoryRepository;
  private final CategorySelector categorySelector;
  private final CategoryFactory categoryFactory;

  private Account newAccount;

  public AddAccountInteractor(AccountRepository accountRepository, AccountFactory accountFactory,
      CategoryRepository categoryRepository, CategorySelector categorySelector,
      CategoryFactory categoryFactory, ThreadExecutor threadExecutor,
      PostExecutionThread postExecutionThread) {
    super(threadExecutor, postExecutionThread);
    this.accountRepository = accountRepository;
    this.accountFactory = accountFactory;
    this.categoryRepository = categoryRepository;
    this.categorySelector = categorySelector;
    this.categoryFactory = categoryFactory;
  }

  public void configure(String accountName, String accountSecret) {
    newAccount = accountFactory.createAccount(accountName, accountSecret);
  }

  @Override
  public Observable<Account> createObservable() {
    if (newAccount == null) {
      throw new IllegalStateException("You must call configure before execute the interactor");
    }

    return categorySelector.getSelectedCategory().flatMap(category -> {
      Category emptyCategory = categoryFactory.createEmptyCategory();
      if (category.equals(emptyCategory)) {
        return accountRepository.add(newAccount);
      }

      return accountRepository.add(newAccount)
          .flatMap(createdAccount -> categoryRepository.addAccount(category, createdAccount)
              .flatMap(categoryWhereAccountJustAdded -> Observable.just(createdAccount)));
    });
  }
}

File: domain/src/main/kotlin/com/arturogutierrez/openticator/domain/account/interactor/AddAccountInteractor.kt
package com.arturogutierrez.openticator.domain.account.interactor

import com.arturogutierrez.openticator.domain.account.interactor.AddAccountInteractor.Params
import com.arturogutierrez.openticator.domain.account.model.Account
import com.arturogutierrez.openticator.domain.account.repository.AccountRepository
import com.arturogutierrez.openticator.domain.category.CategoryFactory
import com.arturogutierrez.openticator.domain.category.CategorySelector
import com.arturogutierrez.openticator.domain.category.repository.CategoryRepository
import com.arturogutierrez.openticator.executor.PostExecutionThread
import com.arturogutierrez.openticator.executor.ThreadExecutor
import com.arturogutierrez.openticator.interactor.SingleUseCase
import io.reactivex.Single

class AddAccountInteractor(private val accountRepository: AccountRepository,
                           private val categoryRepository: CategoryRepository,
                           private val categorySelector: CategorySelector,
                           private val categoryFactory: CategoryFactory,
                           threadExecutor: ThreadExecutor,
                           postExecutionThread: PostExecutionThread) : SingleUseCase<Account, Params>(threadExecutor, postExecutionThread) {

  public override fun buildObservable(params: Params): Single<Account> {
    val newAccount = params.account

    return categorySelector.selectedCategory
        .flatMapSingle { category ->
          val emptyCategory = categoryFactory.createEmptyCategory()
          if (category == emptyCategory) {
            accountRepository.add(newAccount)
          } else {
            accountRepository.add(newAccount)
                .flatMap { createdAccount ->
                  categoryRepository.addAccount(category, newAccount)
                      .flatMap { Single.just(createdAccount) }
                }
          }
        }.firstOrError()
  }

  data class Params(val account: Account)
}
